cluster-autoscaler/cloudprovider/azure/azure_vms_pool.go

```diff
@@ -102,28 +120,260 @@ func (agentPool *VMsPool) MaxSize() int {
+		// otherwise the PUT request will fail
+		if versionedAP.Properties.Mode != nil &&
+			*versionedAP.Properties.Mode == armcontainerservice.AgentPoolModeSystem {
+			systemMode := armcontainerservice.AgentPoolModeSystem
+			requestBody.Properties.Mode = &systemMode
+		}
+
+		// self-hosted CAS will be using Manual scale profile
+		if len(versionedAP.Properties.VirtualMachinesProfile.Scale.Manual) > 0 {
+			// set the count for first manual scale profile to the new target value
+			virtualMachineProfile := *versionedAP.Properties.VirtualMachinesProfile
+			versionedAP.Properties.VirtualMachinesProfile.Scale.Manual[0].Count = to.Int32Ptr(int32(count))
+			requestBody.Properties.VirtualMachinesProfile = &virtualMachineProfile
+			return requestBody, nil
+		}
+		// aks-managed CAS will be using AutoscaleProfile
+		// TODO(wenxuan): support AutoscaleProfile once it become available with the new API release
+	}
+	return armcontainerservice.AgentPool{}, fmt.Errorf("failed to build request body for scale up, agentpool doesn't have valid virtualMachinesProfile")
+}
+
+// scaleUpToCount sets node count for vms agent pool to target value through PUT AP call.
+func (agentPool *VMsPool) scaleUpToCount(count int64) error {
+	agentPool.sizeMutex.Lock()
+	defer agentPool.sizeMutex.Unlock()
+
+	updateCtx, updateCancel := getContextWithCancel()
+	defer updateCancel()
+
+	requestBody, err := agentPool.buildRequestBodyForScaleUp(count)
+	if err != nil {
+		klog.Errorf("Failed to build request body for scale up, error: %s", err)
+		return err
+	}
+
+	klog.V(1).Infof("Scaling up vms pool %s to new target count: %d", agentPool.Name, count)
+	poller, err := agentPool.manager.azClient.agentPoolClient.BeginCreateOrUpdate(
+		updateCtx,
+		agentPool.clusterResourceGroup,
+		agentPool.clusterName,
+		agentPool.Name,
+		requestBody, nil)
+
+	if err != nil {
+		klog.Errorf("Failed to update agentpool %s in cluster %s with error: %v",
+			agentPool.azureRef.Name, agentPool.clusterName, err)
+		return err
+	}
+
+	updateCtx, cancel := getContextWithTimeout(updateRequestContextTimeout)
+	defer cancel()
+
+	if _, err = poller.PollUntilDone(updateCtx, nil); err == nil {
+		// success path
+		klog.Infof("agentPoolClient.BeginCreateOrUpdate for aks cluster %s agentpool %s succeeded", agentPool.clusterName, agentPool.Name)
+		agentPool.curSize = count
+		agentPool.lastSizeRefresh = time.Now()
+		klog.V(6).Infof("setVMsPoolNodeCount: invalidating cache, new size %d", count)
+		agentPool.manager.invalidateCache()
+		return nil
+	}
+
+	klog.Errorf("agentPoolClient.BeginCreateOrUpdate for aks cluster %s agentpool %s failed with error %s",
+		agentPool.clusterName, agentPool.Name, err)
+	return nil
 }
 
 // DeleteNodes extracts the providerIDs from the node spec and
-// delete or deallocate the nodes from the agent pool based on the scale down policy.
+// delete or deallocate the nodes based on the scale down policy of agentpool.
```

Comment:
We don't have a deallocate implementation in the upstream codebase here yet right?

Reply:
We don't have it here in CAS upstream, but AKS-RP is able to deallocate/delete nodes based on property `properties.scaleDownMode` of agentpool. If an VMs pool has `scaleDownMode == Deallocate`, the scale down triggered by CAS will cause RP to deallocate nodes rather than delete.

![image](https://github.com/user-attachments/assets/08588183-2577-4021-a22e-966d6a226ee6)